packages/asset-swapper/src/utils/quote_requestor.ts

```diff
@@ -0,0 +1,100 @@
+import { assetDataUtils, SignedOrder } from '@0x/order-utils';
+import { ERC20AssetData } from '@0x/types';
+import { BigNumber, logUtils } from '@0x/utils';
+import Axios, { AxiosResponse } from 'axios';
+import * as _ from 'lodash';
+
+import { constants } from '../constants';
+import { MarketOperation, RfqtFirmQuoteRequestOpts } from '../types';
+
+/**
+ * Request quotes from RFQ-T providers
+ */
+
+function getTokenAddressOrThrow(assetData: string): string {
+    const decodedAssetData = assetDataUtils.decodeAssetDataOrThrow(assetData);
+    if (decodedAssetData.hasOwnProperty('tokenAddress')) {
+        // type cast necessary here as decodeAssetDataOrThrow returns
+        // an AssetData object, which doesn't necessarily contain a
+        // token address.  (it could possibly be a StaticCallAssetData,
+        // which lacks an address.)  so we'll just assume it's a token
+        // here.  should be safe, with the enclosing guard condition
+        // and subsequent error.
+        // tslint:disable-next-line:no-unnecessary-type-assertion
+        return (decodedAssetData as ERC20AssetData).tokenAddress;
+    } else {
+        throw new Error(`Decoded asset data (${JSON.stringify(decodedAssetData)}) does not contain a token address`);
+    }
+}
+
+export class QuoteRequestor {
+    private readonly _rfqtMakerEndpoints: string[];
+
+    constructor(rfqtMakerEndpoints: string[]) {
+        this._rfqtMakerEndpoints = rfqtMakerEndpoints;
+    }
+
+    public async requestRfqtFirmQuotesAsync(
+        makerAssetData: string,
+        takerAssetData: string,
+        assetFillAmount: BigNumber,
+        marketOperation: MarketOperation,
+        takerApiKey: string,
+        takerAddress: string,
+        options?: Partial<RfqtFirmQuoteRequestOpts>,
+    ): Promise<SignedOrder[]> {
+        const { makerEndpointMaxResponseTimeMs } = _.merge({}, constants.DEFAULT_RFQT_FIRM_QUOTE_REQUEST_OPTS, options);
+
+        const buyToken = getTokenAddressOrThrow(makerAssetData);
+        const sellToken = getTokenAddressOrThrow(takerAssetData);
+
+        // create an array of promises for quote responses, using "undefined"
+        // as a placeholder for failed requests.
+
+        const responsesIfDefined: Array<void | AxiosResponse<SignedOrder>> = await Promise.all(
```

Comment:
A little confused here, I expected this to be `Array<undefined | AxiosResponse<SignedOrder>>`.  What am I missing where this is `void` instead of `undefined`?

Reply:
Great catch.  This is an artifact from when I was doing `map(async endpoint => Axios.get(...).catch(...)`.  Jacob and Lawrence had me change it to be `map(async endpoint => try { Axios.get(...) } catch {...}`.  As an array of potentially rejected promises, `tsc` demanded void.  When I switched to `try{}catch{}`, I introduced a `return undefined;` to the catch handler, which brought everything into alignment on the `undefined` as placeholder notion.  But I neglected to change this `void` to an `undefined`.  Fix is under test...

Reply:
Addressed in https://github.com/0xProject/0x-monorepo/pull/2541/commits/ccc9e18132f0c66d4d97ce6fe94b2759f5290102